power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/dataset.hpp

```diff
@@ -302,20 +302,16 @@ template <dataset_type_tag dataset_type_> class Dataset {
         }
     }
 
-    void add_attribute_buffer(std::string_view component, std::string_view attribute, Data* data) {
-        Idx const idx = find_component(component, true);
-        Buffer& buffer = buffers_[idx];
-        if (!is_columnar(buffer)) {
-            throw DatasetError{"Cannot add attribute buffers to row-based dataset!\n"};
-        }
-        if (std::ranges::find_if(buffer.attributes, [&attribute](auto const& buffer_attribute) {
-                return buffer_attribute.meta_attribute->name == attribute;
-            }) != buffer.attributes.end()) {
-            throw DatasetError{"Cannot have duplicated attribute buffers!\n"};
-        }
-        AttributeBuffer<Data> attribute_buffer{
-            .data = data, .meta_attribute = &dataset_info_.component_info[idx].component->get_attribute(attribute)};
-        buffer.attributes.emplace_back(attribute_buffer);
+    void add_attribute_buffer(std::string_view component, std::string_view attribute, Data* data)
+        requires(!is_indptr_mutable_v<dataset_type>)
+    {
+        add_attribute_buffer_impl(component, attribute, data);
+    }
+
+    void set_attribute_buffer(std::string_view component, std::string_view attribute, Data* data)
+        requires is_indptr_mutable_v<dataset_type>
+    {
+        add_attribute_buffer_impl(component, attribute, data);
```

Comment:
@Jerry-Jinfeng-Guo after offline discussion with Tony, we decided to go with the same behavior between `add_attribute_buffer` and `set_attribute_buffer` (but different entrypoints). The behavior of `set_attribute_buffer` therefore differs from the one of `set_buffer`. The reasoning is as follows:

* For components:
  * the deserializer tells the user via the dataset info that a certain component is present in the serialized data.
  * It is possible to efficiently determine whether that is the case.
  * The user can then only call `set_buffer` for those components that are already present
* For attributes:
  * the deserializer would need to go over the entire dataset to look for components with the map serialization representation to determine whether an attribute is present.
  * this is expensive
  * the deserializer therefore cannot let the user know beforehand which attributes are present
  * `set_attribute_buffer` therefore should only be called if it has not been set yet.
  * this is the same behavior as `add_attribute_buffer`

Reply:
In that case, shouldn't there be something in place to make sure 
> ... therefore should only be called if it has not been set yet.
is actually guarded?

Reply:
it is in https://github.com/PowerGridModel/power-grid-model/pull/711/files#diff-2cc40425b5c3604ed2e2401defd34d90a6fb7050406a81139fc1be14bbfda661R447-R451

```cpp
        if (std::ranges::find_if(buffer.attributes, [&attribute](auto const& buffer_attribute) {
                return buffer_attribute.meta_attribute->name == attribute;
            }) != buffer.attributes.end()) {
            throw DatasetError{"Cannot have duplicated attribute buffers!\n"};
        }
```

Reply:
`set_attribute_buffer` should then be carefully documented, as it behaves different to `set_buffer`, but it make sense due to the expensiveness of searching for specific attributes.

Reply:
Considering renaming it to `set_attribute_buffer_but_not_really`

Reply:
If the renaming to `set_attribute_buffer_but_not_really` doesn't follow through (or even if it does), please add a comment briefly explaining why this doesn't really set anything. I can see one of us very confused in a couple of months about this. A link to this comments would suffice.

Reply:
from the users' perspective, a dataset has all attributes declared for any given component. this keeps the symmetry between row-based and columnar. in a way, we could also rename `add_attribute_buffer` to `set_attribute_buffer`, but then you encounter the exact same naming problem that triggered this discussion in the first place

Reply:
Apart from the docstring that needs addition, I consider this thread resolvable. Could someone involved do it? ;)